### PR TITLE
`QuadFormAndIsom`: final patch before promotion to `src/`

### PR DESCRIPTION
**Pull request by StevellM**
First step of https://github.com/oscar-system/Oscar.jl/issues/4796 . Should not be merged yet, see the TODO list at the bottom. The PR is open to review though.

This PR contains a lot of things as a final patch. I will comment the part of the code which would need a proper review, let me point out here the main changes

* New features
  - Add a new function `splitting`: it generalizes the approach of [BH23] to be able to enumerate conjugacy classes of isometries of any finite order. Based on a naive approach, but works well for our use cases.
  - Add a new **non-exported** function `oscar_genus_representatives`: it is built on top of the `representatives` function on Hecke for `ZZGenus`. The purpose is to be able to have a function to enumerate large genera of definite integer lattices. The idea is to use Kneser's method of neighbor and find a good number of isometry classes using random walks in the associated graph. If after a certain number of vain iterations no new isometry classes are found, we use the machinery of [BH23] to enumerate isomorphism classes of lattices with isometry of prime order.
  - Handle extra "condition" arguments for enumerating isometry classes of lattices with isometry: one can input a minimal/characteristic polynomial for the associated isometries, as well as some invariants of their eigenlattices such as the rank and signatures.
  - For experts, add new optional argument for using databases of definite lattices during genus enumeration, taking care of negative definite lattices with $(-2)$-roots and Galois actions on hermitian genera.
  - For all the previous, add new functions to create a dictionary of eigenlattices conditions, and about Roger's upper bound on center of density for (sphere packing in) definite lattices.

* Bug fix
  - Fix the `fitting_isometry` part of equivariant primitive extensions: previously we were looking at orbits of elements in a certain cosets but the action of the group was not well-defined. Now we shrink the group appropriately to have a honest group action on the coset of isometries, so one can use the `gset` infrastructure.
  - Fix the output check for `_primitive_extensions_generic`: we were checking the genus of the output lattices but in a wrong way, giving rise to output in the wrong genus.
  - Removed the functions `direct_product` and `biproduct`, since there are not technically not mathematically correct. Abit towards the new infrastructure of direct sums, we now use `direct_sum` for getting injections and projections.
  
* Improvements
  - Improve the enumeration of genus symbols for integer lattices in `_find_L`: in a few cases I have tried, we talk about a factor of at least 50 in terms of speed (!!!). The difference is that instead of enumerating genus symbols and keeping the one with the good scale and level, we directly impose the conditions of scale and level in the enumerating function (the infrastructure was there but it was buggy; recently fixed in Hecke). I guess we might also win a lot in allocations here, which is a (big) issue with this project.
  - Improve `splitting_of_hermitian_type`: we only enumerate admissible triples with the correct ranks. It is not a big improvement, since the cases with the wrong ranks are quickly discarded, but we might win in speed on larger framework, and probably some allocations too.
  - Improve the generic function `enumerate_classes_of_lattices_with_isometry` to benefit the new `splitting` function and handling on eigenlattices conditions. Moreover add the possibility not to destroy partial computations. This generic functions should not be used since it is not the best to use in general, but it is good to have it for small examples.
  
* Misc
  - Improve documentation and add more comments in the code whenever it makes sense.
  - Change the formatting to fit into the 80 characters per line (whenever possible), realign function arguments and create intermediate variables to avoid very long lines.
  - Add further sanity checks.
  - Try to improve type stability and allocations.
  
* TODO
After reviewing and before merging it, need to
- [x] Test the code in it current state on a larger scale (lots of computations running on an external machine, hopefully catching any further corner cases)
- [x] Add more doctests for better documentation
- [x] Add more tests to ensure good test coverage (especially regarding latest fixes, improvements and new features)

### CODE REVIEW

File: experimental/QuadFormAndIsom/src/embeddings.jl [hunk 1]
```diff
@@ -364,37 +419,39 @@ function _fitting_isometries(OqfN::AutomorphismGroup{TorQuadModule},
...
   # To summarize, in the general case, we obtain representatives of fitting
   # isometries by identifying CN-conjugate isometries in the coset fNKN.
   if first
     reporb = QQMatrix[solve(basis_matrix(N), basis_matrix(N)*matrix(_fN); side=:left)]
   else
     KNhat, _ = discrep\(kernel(_actN)[1])
+    _CN, _ = centralizer(_actN(stabN)[1], _imN(_fHN))
+    _CN, _ = _actN\_CN
+    CN, _ = discrep\_CN
+    @hassert :ZZLatWithIsom 3 is_normal_subgroup(KNhat, CN)
```
Comment (StevellM) [inline experimental/QuadFormAndIsom/src/embeddings.jl]: Before here the group `CN` was too big and not acting on the coset `fNKN`. This new `CN` is the actual classifying group for the isometries restricting to `_fHN` on the glue groups, and so it acts honestly on `fNKN`, so we use the function `orbits` from the GSet functionalities.

File: experimental/QuadFormAndIsom/src/embeddings.jl [hunk 2]
```diff
@@ -735,15 +799,19 @@ function _primitive_extensions_generic(
...
 
           for b in reporb
             L, fL, graph = _overlattice(phig, HMinD, HNinD, fM, b; same_ambient)
 
-            (!isnothing(q) && genus(L) == G) || (is_even(L) == even) || continue
+            if !isnothing(q) && genus(L) != G
+              continue
+            elseif is_even(L) != even
+              continue
+            end
```
Comment (StevellM) [inline experimental/QuadFormAndIsom/src/embeddings.jl:816]: Previously the test was as `(!isnothing(q) && genus(L) == G) || (is_even(L) == even) || continue` which was wrong because if `q  !== nothing` and `genus(L) != G` we should discard `L`, but we were keeping it if it was even (which did not make sense at all here)

File: experimental/QuadFormAndIsom/src/embeddings.jl [hunk 3]
```diff
@@ -1025,7 +1100,7 @@ function _classes_isomorphic_subgroups(q::TorQuadModule,
     push!(blocks, compose(__f, j))
     push!(ni, ngens(domain(__f)))
   end
-  D, inj, proj = biproduct(domain.(blocks))
+  D, inj, proj = Hecke._biproduct(domain.(blocks))
```
Comment (StevellM) [inline experimental/QuadFormAndIsom/src/embeddings.jl:1105]: Since I should remove the `biproduct` function from user interface in Hecke, I instead call the internal version of it

File: experimental/QuadFormAndIsom/src/embeddings.jl [hunk 4]
```diff
@@ -1063,7 +1138,7 @@ function _classes_isomorphic_subgroups(q::TorQuadModule,
   for lis in Hecke.cartesian_product_iterator(list_can)
     embs = TorQuadModuleMap[l[1] for l in lis]
     embs = TorQuadModuleMap[hom(domain(embs[i]), q, TorQuadModuleElem[blocks[i](domain(blocks[i])(lift(embs[i](a)))) for a in gens(domain(embs[i]))]) for i in 1:length(lis)]
-    H2, _proj = direct_product(domain.(embs)...)
+    H2, _, _proj = Hecke._biproduct(domain.(embs))
```
Comment (StevellM) [inline experimental/QuadFormAndIsom/src/embeddings.jl:1143]: Same as above, `direct_product` should be avoided in the future.

File: experimental/QuadFormAndIsom/src/enumeration.jl [hunk 5]
```diff
@@ -42,47 +48,45 @@ function _find_D(d::T, m::Int, p::IntegerUnion) where T <: IntegerUnion
...
+    return ZZGenus[genus(integer_lattice(; gram=matrix(QQ, 0, 0, [])))]
+  end
+  gen = ZZGenus[]
+  Ip = intersect(max(0, r-nG):min(pG, r), pos) # Possible positive signature
+  In = intersect(max(0, r-pG):min(nG, r), neg) # Possible negative signature
+  for s1 in Ip
+    s2 = r-s1
+    !(s2 in In) && continue
+    # We enumerate all potential genera with the given invariants
+    append!(gen, integer_genera((s1, s2), d; min_scale=s, max_scale=p*l, even))
```
Comment (StevellM) [inline experimental/QuadFormAndIsom/src/enumeration.jl:78]: Big improvement here: we enumerate less genus symbols by imposing the conditions about the scale and level directly into the `integer_genera` function

File: experimental/QuadFormAndIsom/src/enumeration.jl [hunk 6]
```diff
@@ -115,34 +122,58 @@ julia> is_admissible_triple(genus(F), genus(C), genus(Lf), 5)
...
+  # The determinant of C should divide the one of A+B, and the quotient should
+  # be an even power of the prime p
+  _q, _r = divrem(numerator(det(AperpB)), numerator(det(C)))
+  !iszero(_r) && return false
+  ok, _g, _p = is_prime_power_with_data(_q)
+  !ok && return false
+  _p != p && return false
+  g, r = divrem(_g, 2)
+  !iszero(r) && return false
+
```
Comment (StevellM) [inline experimental/QuadFormAndIsom/src/enumeration.jl]: Looks more complicated but eventually that is better. Instead of running over a finite set of value and do a lot of multiplication, we make sure that the determinants satisfy the wanted proprety ($det(A)*det(B) = p^{2g}det(C)$) and then we make sure that $g$ is small enough.

File: experimental/QuadFormAndIsom/src/enumeration.jl [hunk 7]
```diff
@@ -669,7 +921,11 @@ function representatives_of_hermitian_type(G::ZZGenus, chi::Union{ZZPolyRingElem
     @hassert :ZZLatWithIsom 1 is_of_hermitian_type(MfM)
     first && return ZZLatWithIsom[MfM]
 
+    allow_info && println("Enumerate hermitian genus of rank $(rank(H))")
     gr = genus_representatives(H)
+    if is_definite(H)
+      @hassert :ZZLatWithIsom 1 mass(H) == sum(1//automorphism_group_order(HH) for HH in gr)
+    end
```
Comment (StevellM) [inline experimental/QuadFormAndIsom/src/enumeration.jl]: We have observed some weird behavior regarding genus enumeration for hermitian lattices in Hecke. This is temporary until I have finished running the computations on a larger scale (to see if we catch further bugs/corner cases)

File: experimental/QuadFormAndIsom/src/enumeration.jl [hunk 8]
```diff
@@ -716,58 +1023,163 @@ julia> is_of_same_type(Lf, reps[2]^2)
...
+    Lf::ZZLatWithIsom,
+    p::IntegerUnion,
+    b::Int = 0;
+    eiglat_cond::Dict{Int, Vector{Int}}=Dict{Int, Vector{Int}}(),
+    fix_root::Int=-1,
+    genusDB::Union{Nothing, Dict{ZZGenus, Vector{ZZLat}}}=nothing,
+    root_test::Bool=false,
+    check::Bool=true,
+    info_depth::Int=1,
+  )
```
Comment (StevellM) [inline experimental/QuadFormAndIsom/src/enumeration.jl:1052]: This function is at the heart of all the enumeration business and should be reviewed carefully

File: experimental/QuadFormAndIsom/src/enumeration.jl [hunk 9]
```diff
@@ -787,144 +1199,205 @@ julia> splitting_of_prime_power(Lf, 3, 1)
...
 
-  B0 = kernel_lattice(Lf, r)
-  RA = representatives_of_hermitian_type(A0, p, fix_root)
-  is_empty(RA) && return reps
-  RB = splitting_of_pure_mixed_prime_power(B0, p; p_inv, n_inv)
+  # We follow Algorithm 6 of [BH23]: there is a slight mistake in the pseudocode
+  # though, `A_0` and `B_0` should be switched for calling the algorithm
+  # `PrimitiveExtensions`.
+  A0 = kernel_lattice(Lf, r)
+  B0 = kernel_lattice(Lf, n)
```
Comment (StevellM) [inline experimental/QuadFormAndIsom/src/enumeration.jl:1371]: Funny small improvement: the function decomposes `n` as $p^e*q^d$ and later we redo this product instead of using $n$ itself...

File: experimental/QuadFormAndIsom/src/enumeration.jl [hunk 10]
```diff
@@ -988,119 +1470,370 @@ function splitting_of_mixed_prime_power(Lf::ZZLatWithIsom, p::IntegerUnion, b::I
...
+      groups when computing the corresponding $\Phi_k$-kernel sublattices;
+    - if available, one can use any database of genera of definite lattices
+      using the keyword argument `genusDB` (which should be a dictionary whose
+      keys are genus symbols, and the corresponding value is a list of lattices
+      of this genus);
+    - if `root_test` is set to true, the genus enumeration algorithm determines
+      whether any new genus of definite lattices to be enumerated consists of
+      lattices of minimum 2. In such a case, the enumeration is skipped.
+"""
+function splitting(
```
Comment (StevellM) [inline experimental/QuadFormAndIsom/src/enumeration.jl:1590]: New splitting function to be reviewed globally

File: experimental/QuadFormAndIsom/src/enumeration.jl [hunk 11]
```diff
@@ -1111,67 +1844,438 @@ end
 
 ###############################################################################
 #
-#  Testing functions
+#  Helper for managing conditions on isometries to enumerate
```
Comment (StevellM) [inline experimental/QuadFormAndIsom/src/enumeration.jl:1896]: New section about handling conditions on eigenlattices and minimal/characteristic polynomials (as well as understanding the restrictions on the powers of the isometries)

File: experimental/QuadFormAndIsom/src/enumeration.jl [hunk 12]
```diff
@@ -1111,67 +1844,438 @@ end
...
+      V[m] = max.([-1, -1, -1], t)
     end
-    D[p^k*q^j] = Dn
   end
+  return V
+end
+
+###############################################################################
+#
+#  Enhanced genus enumeration
```
Comment (StevellM) [inline experimental/QuadFormAndIsom/src/enumeration.jl:2013]: To be reviewed globally: add Roger's upper bound and `oscar_genus_representatives`.

File: experimental/QuadFormAndIsom/src/embeddings.jl [hunk 13]
```diff
@@ -290,64 +346,63 @@ function _possible_glue_orders(qM::TorQuadModule,
...
-                             phig::TorQuadModuleMap,
-                             fHM::TorQuadModuleMap,
-                             discrep::GAPGroupHomomorphism,
-                             stabN::AutomorphismGroup{TorQuadModule},
-                             N::ZZLat,
-                             first::Bool)
+# We look for representatives of orbits of isometries on a lattice $N$ which
+# fit in a given gluing context, i.e. for which a given glue map is
+# equivariant. These isometries all live in a same coset, and we identity two
+# isometries in Let $N$ be the associated lattice.
```
Comment (simonbrandhorst) [inline experimental/QuadFormAndIsom/src/embeddings.jl]: Looking a this comment. I still don't know what the function actually computes.

File: experimental/QuadFormAndIsom/src/embeddings.jl [hunk 14]
```diff
@@ -364,37 +419,39 @@ function _fitting_isometries(OqfN::AutomorphismGroup{TorQuadModule},
   # _fHN is the coset _fN*KN where KN is the preimage by discrep of the kernel
   # of _imN. Now, inside this coset, some isometries could still give rise to
   # isomorphic equivariant primitive extensions for our classifying group. Thus
-  # we need to identity isometries which are conjugate by an isometry of our
+  # we need to identify isometries which are conjugate by an isometry of our
   # classifying group stabilizing HN (otherwise it does not make sense). This
```
Comment (simonbrandhorst) [inline experimental/QuadFormAndIsom/src/embeddings.jl:425]: what is "the classifying group"?

File: experimental/QuadFormAndIsom/src/embeddings.jl (hunk 1 above)
Comment (simonbrandhorst) [reply experimental/QuadFormAndIsom/src/embeddings.jl]: Can't comment here, as long as I do not know what this function actually does.

File: experimental/QuadFormAndIsom/src/embeddings.jl (hunk 2 above)
Comment (simonbrandhorst) [reply experimental/QuadFormAndIsom/src/embeddings.jl:816]: Agreed.

File: experimental/QuadFormAndIsom/src/enumeration.jl (hunk 6 above)
Comment (simonbrandhorst) [reply experimental/QuadFormAndIsom/src/enumeration.jl]: o.k.

File: experimental/QuadFormAndIsom/src/enumeration.jl (hunk 5 above)
Comment (simonbrandhorst) [reply experimental/QuadFormAndIsom/src/enumeration.jl:78]: I finally understand why that was a bottleneck :D.

File: experimental/QuadFormAndIsom/src/enumeration.jl [hunk 15]
```diff
@@ -151,15 +182,23 @@ function is_admissible_triple(A::ZZGenus, B::ZZGenus, C::ZZGenus, p::IntegerUnio
     return false
   end
```
Comment (simonbrandhorst) [inline experimental/QuadFormAndIsom/src/enumeration.jl:192]: How sure are you about the excess condition? Is it equivalent to being rationally equivalent?
(We also have dedicated functions for this.)

File: experimental/QuadFormAndIsom/src/enumeration.jl [hunk 16]
```diff
@@ -678,17 +934,68 @@ function representatives_of_hermitian_type(G::ZZGenus, chi::Union{ZZPolyRingElem
...
-Note that $e$ can be `0`.
+The integer `b` can be set to be `0` or `1`, depending on whether one allows
+`Lf` to be among the outputs or not. For instance, setting `b = 1`
+would enforce every $(M, g)$ in output to satisfy that $g$ has order $p*m$.
+
+For every $(M, g)$ in output, one may decide on the rank `r`, the positive
+signature `p` and the negative signature `n` of the eigenlattices of $(M, g)$
+using the keyword argument `eiglat_cond`. It should consist of a dictionary
+where each key is a divisor of $p*m$, and the corresponding value is a tuple
+`(r, p, n)` of integers.
```
Comment (simonbrandhorst) [inline experimental/QuadFormAndIsom/src/enumeration.jl]: Looks redundant to me since `r=p+n`.

File: experimental/QuadFormAndIsom/src/enumeration.jl [hunk 17]
```diff
@@ -678,17 +934,68 @@ function representatives_of_hermitian_type(G::ZZGenus, chi::Union{ZZPolyRingElem
...
+    - by setting the value of `fix_root` to a certain integer $k$, the function
+      only compute one generator for every conjugacy classes of finite cyclic
+      groups when computing the corresponding $\Phi_k$-kernel sublattices;
+    - if available, one can use any database of genera of definite lattices
+      using the keyword argument `genusDB` (which should be a dictionary whose
+      keys are genus symbols, and the corresponding value is a list of lattices
+      of this genus);
+    - if `root_test` is set to true, the genus enumeration algorithm determines
+      whether any new genus of definite lattices to be enumerated consists of
+      lattices of minimum 2. In such a case, the enumeration is skipped.
```
Comment (simonbrandhorst) [inline experimental/QuadFormAndIsom/src/enumeration.jl]:
```suggestion
    When using this function, one can use some extra keyword arguments
    which are carried along the computation:
    - by setting the value of `fix_root` to a certain integer $k$, the function
      only computes one generator for every conjugacy class of finite cyclic
      group;
    - if available, one can use any database of genera of definite lattices
      using the keyword argument `genusDB` (which should be a dictionary whose
      keys are genus symbols, and the corresponding value is a list of lattices
      of this genus);
    - if `root_test` is set to true, the genus enumeration algorithm determines
      whether any new genus of definite lattices to be enumerated consists only of
      lattices of minimum 2. In such a case, the enumeration is skipped.
```

File: experimental/QuadFormAndIsom/src/enumeration.jl (hunk 8 above)
Comment (simonbrandhorst) [reply experimental/QuadFormAndIsom/src/enumeration.jl:1052]: Done. Looked good to me.

File: experimental/QuadFormAndIsom/src/enumeration.jl (hunk 12 above)
Comment (simonbrandhorst) [reply experimental/QuadFormAndIsom/src/enumeration.jl:2013]: Okay. I did not check for typos in bn.

File: experimental/QuadFormAndIsom/src/embeddings.jl (hunk 13 above)
Comment (StevellM) [reply experimental/QuadFormAndIsom/src/embeddings.jl]: Looks like I deleted parts of the comment. You have a primitive extension $M\oplus N\subseteq L$, with $N$ negative definite. We fix an isometry $f$ of $M$ and we want to determine all isometries $g$ of $N$ such that $f\oplus g$ preserves $L$. However, we do not want all such isometries $g$  but only representatives for certain orbits of such (usually up to the action of $O(L, N)$ )

File: experimental/QuadFormAndIsom/src/enumeration.jl (hunk 15 above)
Comment (StevellM) [reply experimental/QuadFormAndIsom/src/enumeration.jl:192]: This was originally in your code. I guess we checked it together at some point, at the very beginning of the project, but I could look at it again if necessary.

File: experimental/QuadFormAndIsom/src/enumeration.jl (hunk 16 above)
Comment (StevellM) [reply experimental/QuadFormAndIsom/src/enumeration.jl]: I should probably mention that they could be $-1$ . For instance, you could mention $(-1, 2, -1)$ if you only know your eigenlattice should have positive signature 2, but you do not specify the rank

File: experimental/QuadFormAndIsom/src/enumeration.jl (hunk 15 above)
Comment (simonbrandhorst) [reply experimental/QuadFormAndIsom/src/enumeration.jl:192]: Now the question is, do I trust myself? :D

File: experimental/QuadFormAndIsom/src/embeddings.jl [hunk 18]
```diff
@@ -1237,10 +1348,15 @@ $L'$ isometric to $M$, and $N'$ is the orthogonal complement of $M'$ in $L'$.
 If `T == false`, then $V$ will always be the empty list. If `T == true`, then
 the content of $V$ depends on the value of the symbol `classification`. There
 are 4 possibilities:
```
Comment (fingolfin) [inline experimental/QuadFormAndIsom/src/embeddings.jl]:
```suggestion
are four possibilities:
```

File: experimental/QuadFormAndIsom/src/enumeration.jl (hunk 15 above)
Comment (StevellM) [reply experimental/QuadFormAndIsom/src/enumeration.jl:192]: At that point, we know that both symbols represent nonsingular forms, and the determinants of the symbols differ by a square, so to compare them rationally, we need to compare their $p$-excess (Conway--Sloane at least claim it; see Theoren 3 in Chapter 5 section 5.1 --- page 372). Apparently, it is the same as comparing their Hasse invariant, which seems to be more standard.

File: experimental/QuadFormAndIsom/src/enumeration.jl (hunk 12 above)
Comment (StevellM) [reply experimental/QuadFormAndIsom/src/enumeration.jl:2013]: I just double-checked and it agrees with Conway--Sloane table

File: experimental/QuadFormAndIsom/src/embeddings.jl (hunk 1 above)
Comment (StevellM) [reply experimental/QuadFormAndIsom/src/embeddings.jl]: Please let me know if the new comment helps

File: experimental/QuadFormAndIsom/src/embeddings.jl (hunk 14 above)
Comment (StevellM) [reply experimental/QuadFormAndIsom/src/embeddings.jl:425]: It is vague.. when you classify (equivariant) primitive extensions, you compute some double cosets with respect to subgroup of isometries of each lattice. The "classifying group" is the representation on the glue module of the stabilizer of the primitive extension in one of such group of isometries.

File: experimental/QuadFormAndIsom/src/embeddings.jl (hunk 18 above)
Comment (StevellM) [reply experimental/QuadFormAndIsom/src/embeddings.jl]: Thanks! I have spread the suggestion to the other docstrings

File: experimental/QuadFormAndIsom/src/embeddings.jl [hunk 19]
```diff
@@ -290,64 +346,66 @@ function _possible_glue_orders(qM::TorQuadModule,
...
-                             phig::TorQuadModuleMap,
-                             fHM::TorQuadModuleMap,
-                             discrep::GAPGroupHomomorphism,
-                             stabN::AutomorphismGroup{TorQuadModule},
-                             N::ZZLat,
-                             first::Bool)
+# We are given a primitive extension $M\oplus N \subseteq L$ with $N$
+# negative definite. We assume that we have fixed an isometry $f\in O(M)$.
+# We want to determine isometries $g$ of $N$ such that $f\oplus g$ preserves
+# the lattice $L$ (called them "fitting isometries").
```
Comment (simonbrandhorst) [inline experimental/QuadFormAndIsom/src/embeddings.jl]:
```suggestion
# the lattice $L$ (call them "fitting isometries").
```

File: experimental/QuadFormAndIsom/src/embeddings.jl [hunk 20]
```diff
@@ -290,64 +346,66 @@ function _possible_glue_orders(qM::TorQuadModule,
...
-                             fHM::TorQuadModuleMap,
-                             discrep::GAPGroupHomomorphism,
-                             stabN::AutomorphismGroup{TorQuadModule},
-                             N::ZZLat,
-                             first::Bool)
+# We are given a primitive extension $M\oplus N \subseteq L$ with $N$
+# negative definite. We assume that we have fixed an isometry $f\in O(M)$.
+# We want to determine isometries $g$ of $N$ such that $f\oplus g$ preserves
+# the lattice $L$ (call them "fitting isometries").
+# * `OqfN` here is the image of $O(N)\to O(qN)$ where `qN` is the discriminant
```
Comment (simonbrandhorst) [inline experimental/QuadFormAndIsom/src/embeddings.jl]: is f involved in any way here? The notation suggests it.

File: experimental/QuadFormAndIsom/src/embeddings.jl [hunk 21]
```diff
@@ -364,37 +422,39 @@ function _fitting_isometries(OqfN::AutomorphismGroup{TorQuadModule},
   # _fHN is the coset _fN*KN where KN is the preimage by discrep of the kernel
   # of _imN. Now, inside this coset, some isometries could still give rise to
   # isomorphic equivariant primitive extensions for our classifying group. Thus
-  # we need to identity isometries which are conjugate by an isometry of our
+  # we need to identify isometries which are conjugate by an isometry of our
   # classifying group stabilizing HN (otherwise it does not make sense). This
-  # group of isometries is exactly the preimage of stabN by discrep, which we
-  # call CN here.
+  # group of isometries is exactly the preimage by discrep of the centralizer
+  # in stabN of _fN, which we call CN here.
```
Comment (simonbrandhorst) [inline experimental/QuadFormAndIsom/src/embeddings.jl]: what is `fN`? I thought that is what you want to compute?

File: experimental/QuadFormAndIsom/src/embeddings.jl [hunk 22]
```diff
@@ -364,37 +422,39 @@ function _fitting_isometries(OqfN::AutomorphismGroup{TorQuadModule},
...
   if first
     reporb = QQMatrix[solve(basis_matrix(N), basis_matrix(N)*matrix(_fN); side=:left)]
   else
     KNhat, _ = discrep\(kernel(_actN)[1])
+    _CN, _ = centralizer(_actN(stabN)[1], _imN(_fHN))
+    _CN, _ = _actN\_CN
+    CN, _ = discrep\_CN
+    @hassert :ZZLatWithIsom 3 is_normal_subgroup(KNhat, CN)
+    @hassert :ZZLatWithIsom 3 all(g -> g * _fN == _fN * g, gens(CN))
     fNKN = _fN*KNhat
```
Comment (simonbrandhorst) [inline experimental/QuadFormAndIsom/src/embeddings.jl]: As far as I can see `_fN` is not defined in the body of the function. Is this code tested? It should error. Or am I missing something?

File: experimental/QuadFormAndIsom/src/embeddings.jl (hunk 1 above)
Comment (simonbrandhorst) [reply experimental/QuadFormAndIsom/src/embeddings.jl]: A little. But I still don't understand what kind of cosets are actually computed (and why). 
Please a mathematical definition of the output of the function.

File: experimental/QuadFormAndIsom/src/embeddings.jl (hunk 20 above)
Comment (StevellM) [reply experimental/QuadFormAndIsom/src/embeddings.jl]: Yes, as I have written below, the input `fHM` of the function is the representation of `f` on the glue subgroup `HM` inside the discriminant group of `M`. We do not see `f` directly, by just how it acts on the gluing.

File: experimental/QuadFormAndIsom/src/embeddings.jl (hunk 21 above)
Comment (StevellM) [reply experimental/QuadFormAndIsom/src/embeddings.jl]: You have computed one fitting isometry `_fN` above, in line 414. The way we did it is that we know that any such isometry must coincide with `fHM` on the gluing. The corresponding action on `HN` has been computed as `_fHN`. In line 428, I should have written `_fHN`. I will rework again a bit the comment and keep "f" for isometries on the side of the lattice $M$, and use "g" for the isometries on the side of $N$.

File: experimental/QuadFormAndIsom/src/embeddings.jl (hunk 1 above)
Comment (StevellM) [reply experimental/QuadFormAndIsom/src/embeddings.jl]: When are given a primitive extension $M\oplus N \subseteq L$ and a fixed isometry $f$ of $M$. We suppose that $N$ is negative definite, and we let $\gamma\colon H_M\to H_N$ be the glue map associated to the previous primitive extension. We assume moreover that $D_f$, the isometry of the discriminant group $D_M$ induced by $f$ preserves the glue domain $H_M$.
 We wish to determine all isometries $g$ of $N$ such that $f\oplus g$ defines an isometry of $L$.
 
 For this, we let $f_{H_M}$ be the isometry induced by $f$ on the glue domain: we know that any isometry of $N$ as we want (one line before) should preserve $H_N$ and restrict to $g_{H_N} := \gamma f_{H_M}\gamma^{-1}$ on that module. All such isometries live in a coset of $O(N)$. To compute this coset, we compute one "fitting isometry", i.e. one isometry $g$ of $N$ that restricts to $g_{H_N}$ on the glue. Then any other such isometry lies in $gK_N$ where $K_N$ is the subgroup consisting of isometries which restrict to the identity on $H_N$. In the code, we define this coset in line 441 as `fNKN`.
 
 Now, we are working in a context where the glue map $\gamma$ represents a certain double coset of glue maps. There are two "classifying groups" $O_M$ and $O_N$ such that $\gamma$ represents a coset in $O_M\backslash O(H_N)/O_N$. In input of the algorithm, `stabN` is the stabilizer of $H_N$ in the image of $O_N$ in $O(D_N)$. The algorithm returns a complete set of representatives for conjugacy classes of isometries in $gK_N$ (aka `fNKN` in the code) under the action of $O_N$.

File: experimental/QuadFormAndIsom/src/embeddings.jl (hunk 22 above)
Comment (StevellM) [reply experimental/QuadFormAndIsom/src/embeddings.jl]: You probably need to expand the code, it is defined in line 414. Note that the code is indeed tested and it does not seem to error.

File: experimental/QuadFormAndIsom/src/embeddings.jl (hunk 20 above)
Comment (simonbrandhorst) [reply experimental/QuadFormAndIsom/src/embeddings.jl]: I mean  " `OqfN` here is the image of $O(N)\to O(qN)$ " 
The image of the map $O(N)\to O(qN)$ is independent of $f$.

File: experimental/QuadFormAndIsom/src/embeddings.jl (hunk 1 above)
Comment (simonbrandhorst) [reply experimental/QuadFormAndIsom/src/embeddings.jl]: The first two paragraphs are clear to me. The third one not. 

In the first two parts you assume that gamma is fixed. In the 3. paragraph you choose 
$\gamma$ in a "certain double coset".  How does $O_N$ act on $gK_n$?

File: experimental/QuadFormAndIsom/src/embeddings.jl (hunk 20 above)
Comment (StevellM) [reply experimental/QuadFormAndIsom/src/embeddings.jl]: Ah yes, this is just a bad variable name. I just re-use the name from the only function that calls this internal. I will change it.

File: experimental/QuadFormAndIsom/src/embeddings.jl (hunk 1 above)
Comment (StevellM) [reply experimental/QuadFormAndIsom/src/embeddings.jl]: This internal is only called by the global generic function on primitive extension. At the stage of the algorithm where this function is called, the glue map `gamma` represents a certain double coset. This internal does not see it, but for the sake of the global algorithm, we need to keep track of the "classifying group" to avoid overcounting these "fitting isométries". So yes `gamma` is fixed, I just recalled where the group came from. And the way $O_N$ acts: actually, we look at the isometries in $O_N$ preserving the glue group $H_N$ and then that stabilizer (called `CN` in the code) acts  **by conjugacy** on the coset `fNKN`.

Comment (simonbrandhorst) [reply experimental/QuadFormAndIsom/src/embeddings.jl]: O.K. thanks for the explanations.

**Review by simonbrandhorst — APPROVED** [review]

**Comment by StevellM** [comment]
This is ready to go on my side; not sure whether the tests failing have anything to do with my PR.
